File: solve.py
import math
from numba import guvectorize, vectorize, boolean, float64
import numpy as np


@guvectorize("(n)->(n)", nopython=True)
def section_properties(stringer_locs, res):
    str_count = stringer_locs.shape[0]
    props = np.zeros((4 + str_count, 7))

    # Seed 4 parts that are always there
    ## rear spar
    props[0, 0] = 0.0937500000000  # Area
    props[0, 1] = 3.9375000000000  # y
    props[0, 2] = 0.8750000000000  # z
    props[0, 3] = 0.0043945312500  # Iyy*
    props[0, 4] = 0.0001220703125  # Izz*
    props[0, 5] = 0.0000000000000  # Iyz*
    props[0, 6] = 1926.00000000e3  # modulus of elasticity

    ## top skin
    props[1, 0] = 0.12525
    props[1, 1] = 2
    props[1, 2] = 1.354
    props[1, 3] = 0.0007491
    props[1, 4] = 0.16706
    props[1, 5] = 0
    props[1, 6] = 2036e3

    ## front spar
    props[2, 0] = 0.15625
    props[2, 1] = 0.0625
    props[2, 2] = 0.625
    props[2, 3] = 0.02034505208
    props[2, 4] = 0.0002034505208
    props[2, 5] = 0
    props[2, 6] = 1926e3

    ## bottom skin
    props[3, 0] = 0.12597
    props[3, 1] = 2
    props[3, 2] = 0.23438
    props[3, 3] = 0.00263426
    props[3, 4] = 0.16796
    props[3, 5] = 0.0209929
    props[3, 6] = 2036e3

    # Stringer properties
    props[4:, 0] = 0.05625
    props[4:, 1:2] = 0.0000203451  # Iyy & Izz are identical Iyz = 0
    props[4:, 6] = 2.010e6

    for i in range(stringer_locs.shape[0]):
        if stringer_locs[i] < 4:
            props[1 + i, 1] = (
                15 * math.cos(0.066865794705 * stringer_locs[i] + 1.43706473738) + 2
            )
            props[1 + i, 1] = (
                15 * math.sin(0.066865794705 * stringer_locs[i] + 1.43706473738)
                - 13.6160687473
            )
        else:
            props[1 + i, 1] = 8 - stringer_locs[i]
            props[1 + i, 2] = props[1 + i, 1] / 8

    y_bar = np.sum(props[:, 6] * props[:, 0] * props[:, 1]) / np.sum(
        props[:, 6] * props[:, 0]
    )
    z_bar = np.sum(props[:, 6] * props[:, 0] * props[:, 2]) / np.sum(
        props[:, 6] * props[:, 0]
    )
    # Iyy
    res[0] = np.sum(props[:, 6] * (props[:, 3] + props[:, 0] * (props[:, 2] - z_bar)))
    # Izz
    res[1] = np.sum(props[:, 6] * (props[:, 4] + props[:, 0] * (props[:, 1] - y_bar)))
    # Iyz
    res[2] = np.sum(
        props[:, 6]
        * (props[:, 5] + props[:, 0] * (props[:, 1] - y_bar) * (props[:, 2] - z_bar))
    )
    res[3] = y_bar
    res[4] = z_bar


@vectorize([float64(float64, float64, float64, float64)], nopython=True)
def deflection_y(i_yy, i_zz, i_yz, load):
    return (i_yz / (i_yy * i_zz - i_yz**2)) * (load - 5) * -32180.83333


@vectorize([float64(float64, float64, float64, float64)], nopython=True)
def deflection_z(i_yy, i_zz, i_yz, load):
    return -(i_zz / (i_yy * i_zz - i_yz**2)) * (load - 5) * -32180.83333


@vectorize([float64(float64, float64)], nopython=True)
def shear_flow(load, y_bar):
    return (load * (y_bar - 1) + 5 * y_bar - 10) / 4.136


@vectorize(nopython=True)
def twist(load, y_bar):
    q = shear_flow(load, y_bar)
    return 4.43525e-6 * q**2 + 0.0334937 * q


@vectorize([float64(float64, float64, float64)], nopython=True)
def stress_shear(load, y_bar, thickness):
    return (load * (y_bar - 1) + 5 * y_bar - 10) / (4.136 * thickness)


@vectorize(
    nopython=True,
)
def stress_normal(
    mod_elastic,
    x,
    y,
    z,
    load,
    i_yy,
    i_zz,
    i_yz,
):
    return -y * ((load - 5) * (x - 45.75) * i_yz / (i_yy * i_zz - i_yz**2)) + z * (
        (load - 5) * (x - 45.75) * i_zz / (i_yy * i_zz - i_yz**2)
    )


@vectorize(
    [
        boolean(
            float64,
            float64,
            float64,
            float64,
            float64,
            float64,
            float64,
            float64,
            float64,
            float64,
        )
    ],
    nopython=True,
)
def failed(x, y, z, mod_elastic, load, i_yy, i_zz, i_yz, y_bar, thickness):
    shear = stress_shear(load, y_bar, thickness)
    normal = stress_normal(mod_elastic, x, y, z, load, i_yy, i_zz, i_yz)

    f_1 = -0.038
    f_11 = 0.0096
    f_66 = 0.44

    if thickness < 1 / 8:
        f_1 = -0.054
        f_11 = 0.0072

    if thickness < 1 / 16:
        f_1 = -0.048
        f_11 = 0.0078
        f_66 = 0.31

    return f_1 * normal + f_11 * normal**2 + f_66 * shear**2 >= 1


@guvectorize("(n)->()")
def count_stacked(places, out):
    _, counts = np.unique(places, return_counts=True)
    out[0] = np.sum(counts) - places.shape[0]


@guvectorize("(n)->()")
def count_adjacent(places, out):
    unique = np.unique(places)
    dup = np.roll(unique, 1)
    out[0] = np.count_nonzero(unique - dup <= 1 / 8)


@vectorize
def final_score(n_stringer, n_ribs, n_stacked, n_adj, load, def_q, def_max, twist):
    weight = 0.4369 + n_ribs * 0.00213 + n_stringer * 0.0146
    design = 100 * (
        0.6 * 8 / (n_stringer + 1)
        + 0.3 * 15 / (n_ribs + 1)
        - (n_stacked / 8 + n_adj / 8)
    )
    # removed twist becaues it was causing more issues than it was worth to keep in
    perf = 0.5 * load / weight + 0.5 / def_q + 0.05 * (load / def_max) - 10 * weight
    return design + perf

File: julia/solve.jl
function section_properties(stringer_locs)
    vec_len = size(stringer_locs, 1) + 4
    A = zeros(vec_len)
    y = zeros(vec_len)
    z = zeros(vec_len)
    Iyy = zeros(vec_len)
    Izz = zeros(vec_len)
    Iyz = zeros(vec_len)
    E = zeros(vec_len)

    ## rear spar
    A[1] = 0.09375
    y[1] = 3.9375
    z[1] = 0.875
    Iyy[1] = 0.00439453125
    Izz[1] = 0.0001220703125
    Iyz[1] = 0
    E[1] = 1926e3

    ## top skin
    A[2] = 0.12525
    y[2] = 2
    z[2] = 1.354
    Iyy[2] = 0.0007491
    Izz[2] = 0.16706
    Iyz[2] = 0
    E[2] = 2036e3

    ## front spar
    A[3] = 0.15625
    y[3] = 0.0625
    z[3] = 0.625
    Iyy[3] = 0.02034505208
    Izz[3] = 0.0002034505208
    Iyz[3] = 0
    E[3] = 1926e3

    ## bottom skin
    A[4] = 0.12597
    y[4] = 2
    z[4] = 0.23438
    Iyy[4] = 0.00263426
    Izz[4] = 0.16796
    Iyz[4] = 0.0209929
    E[4] = 2036e3

    for (i, str_pos) in pairs(stringer_locs)
        if str_pos < 4
            y[4+i] = 15 * cos(0.066865794705 * str_pos + 1.43706473738) + 2
            z[4+i] = 15 * sin(0.066865794705 * str_pos + 1.43706473738) - 13.6160687473
        else
            y[4+i] = 8 - str_pos
            z[4+i] = y[4+i] / 8
        end
    end

    y_bar = esum(E .* A .* y) / sum(E .* A)
    z_bar = sum(E .* A .* z) / sum(E .* A)
    return (
        y_bar=y_bar,
        z_bar=z_bar,
        Iyy=sum(E .* (Iyy .+ A .* (z .- z_bar) .^ 2)),
        Izz=sum(E .* (Izz .+ A .* (y .- y_bar) .^ 2)),
        Iyz=sum(E .* (Iyz .+ A .* (y .- y_bar) .* (z .- z_bar))),
    )
end


function deflection_y(mod_elastic, i_yy, i_zz, i_yz, load)
    return (i_yy / (mod_elastic .* (i_yy .* i_zz .- i_yz .^ 2))) .* (load .- 5) .* -32180.8333333
end

function deflection_z(mod_elastic, i_yy, i_zz, i_yz, load)
    return (i_zz / (mod_elastic .* (i_yy .* i_zz .- i_yz .^ 2))) .* (load .- 5) .* -32180.8333333
end

function shear_flow(load, y_bar)
    return (load .* (y_bar .- 1) .+ 5 .* y_bar - 10) ./ 4.136
end

function torsion(x, load, y_bar)
    return 0.00018401233 .* x .* shear_flow(load, y_bar)
end

function stress_shear(load, y_bar, thickness)
    return (load .* (y_bar .- 1) .+ 5 .* y_bar .- 10) ./ (4.136 .* thickness)
end

function stress_normal(x, y, z, load, i_yy, i_zz, i_yz)
    return .-y .* ((load .- 5) .* (x .- 45.75) .* i_yy ./ (i_yy .* i_zz .- i_yz .^ 2)) .+ z .* (
        (load .- 5) .* (x .- 45.75) .* i_zz ./ (i_yy .* i_zz .- i_yz .^ 2)
    )
end

function failed()
    shear = stress_shear(load, y_bar, thickness)
    normal = stress_normal(x, y, z, load, i_yy, i_zz, i_yz)

    f_1 = -0.038
    f_11 = 0.0096
    f_66 = 0.44

    if thickness < 1 / 8
        f_1 = -0.054
        f_11 = 0.0072
    end

    if thickness < 1 / 16
        f_1 = -0.048
        f_11 = 0.0078
        f_66 = 0.31
    end

    return f_1 .* normal .+ f_11 .* normal .^ 2 .+ f_66 .* shear .^ 2 >= 1
end

function count_stacked(stringer_locs)
    sum([count(==(loc), stringer_locs) for loc in unique(stringer_locs)]) - size(stringer_locs, 1)
end

function count_adjacent(stringer_locs)
    uniq = unique(stringer_locs)
    dup = circshift(uniq, 1)
    return sum(uniq - dup .<= 1 / 8)
end

function final_score(n_stringer, n_ribs, n_stacked, n_adj, load, def_q, def_max, twist)
    weight = 0.4369 .+ n_ribs .* 0.00213 .+ n_stringer .* 0.0146
    design = 100 .* (
        0.6 .* 8 ./ (n_stringer .+ 1)
        .+
        0.3 .* 15 ./ (n_ribs .+ 1)
        .-
        (n_stacked ./ 8 .+ n_adj ./ 8)
    )
    perf = (
        0.5 .* load ./ weight
        .+
        0.5 ./ def_q
        .+
        0.05 .* (load ./ def_max .+ load ./ twist)
        .-
        10 .* weight
    )
    return design .+ perf
end